src/Servers/Kestrel/Core/src/Internal/Infrastructure/HttpUtilities.cs

```diff
@@ -86,52 +88,65 @@ private static unsafe ulong GetMaskAsLong(byte[] bytes)
         }
 
         // The same as GetAsciiStringNonNullCharacters but throws BadRequest
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static unsafe string GetHeaderName(this ReadOnlySpan<byte> span)
         {
             if (span.IsEmpty)
             {
                 return string.Empty;
             }
 
-            var asciiString = new string('\0', span.Length);
+            fixed (byte* source = &MemoryMarshal.GetReference(span))
```

Comment:
Does this perform better than `fixed (byte* buffer = span)` used previously?

Reply:
Yes (it safes one length == 0 check).
`GetPinnableReference` needs to check for empty span, this check is already done above, so we are sure that we have a non-empty span, so can skip the check for empty by using `MemoryMarshal.GetReference`.

---

I'll address the other comments tomorrow.